File: apps/php-slim/src/Support/Actor.php
<?php

declare(strict_types=1);

namespace DarkStore\Support;

use Psr\Http\Message\ServerRequestInterface;

/**
 * The caller, reduced to the attributes a targeting rule would plausibly
 * want. Built once per request from forwarded auth headers.
 */
final readonly class Actor
{
    public function __construct(
        public string $key,
        public ?string $email,
        public string $plan,
        public string $country,
    ) {
    }

    public static function fromRequest(ServerRequestInterface $request): self
    {
        $header = static function (string $name, ?string $fallback) use ($request): ?string {
            $value = $request->getHeaderLine($name);

            return $value === '' ? $fallback : $value;
        };

        return new self(
            key: $header('X-User-Key', 'anonymous') ?? 'anonymous',
            email: $header('X-User-Email', null),
            plan: $header('X-User-Plan', 'free') ?? 'free',
            country: $header('X-User-Country', 'US') ?? 'US',
        );
    }
}

File: apps/dotnet-api/Actor.cs
namespace DarkStore.Api;

/// <summary>
/// The caller, reduced to the attributes a targeting rule would plausibly
/// want. Built once per request from forwarded auth headers.
/// </summary>
public sealed record Actor(string Key, string? Email, string Plan, string Country)
{
    public static Actor FromRequest(HttpRequest request)
    {
        string Header(string name, string fallback) =>
            request.Headers.TryGetValue(name, out var values) && values.Count > 0
                ? values[0] ?? fallback
                : fallback;

        return new Actor(
            Key: Header("X-User-Key", "anonymous"),
            Email: request.Headers.TryGetValue("X-User-Email", out var email) ? email[0] : null,
            Plan: Header("X-User-Plan", "free"),
            Country: Header("X-User-Country", "US"));
    }
}
